internals/daemon/daemon.go

```diff
@@ -717,21 +717,79 @@ func (d *Daemon) doReboot(sigCh chan<- os.Signal, waitTimeout time.Duration) err
 	return fmt.Errorf("expected reboot did not happen")
 }
 
-var shutdownMsg = "reboot scheduled to update the system"
+const rebootMsg = "reboot scheduled to update the system"
 
-func rebootImpl(rebootDelay time.Duration) error {
+var rebootHandler = commandReboot
+
+// SetSyscallReboot replaces the default command-based reboot
+// with a direct Linux kernel syscall based implementation.
+func SetSyscallReboot() {
+	rebootHandler = syscallReboot
+}
+
+// commandReboot assumes a userspace shutdown command exists.
+func commandReboot(rebootDelay time.Duration) error {
 	if rebootDelay < 0 {
 		rebootDelay = 0
 	}
 	mins := int64(rebootDelay / time.Minute)
-	cmd := exec.Command("shutdown", "-r", fmt.Sprintf("+%d", mins), shutdownMsg)
+	cmd := exec.Command("shutdown", "-r", fmt.Sprintf("+%d", mins), rebootMsg)
 	if out, err := cmd.CombinedOutput(); err != nil {
 		return osutil.OutputErr(out, err)
 	}
 	return nil
 }
 
-var reboot = rebootImpl
+var (
+	// checkCapSysBoot returns nil if the system has the correct
+	// permissions to issue a reboot request to the Linux kernel
+	checkCapSysBoot = func() error {
```

Comment:
The way we typically do this is to define a real function, and then define a variable that can be overriden, so code inside the export_test.go file can replace it, using a function prefixed with `Fake` and that returns a restoring function named `restore`.

With that said, all of this logic feels like it's adding significant complexity to the code with no benefit, because the test is testing the test code. In other words, all of this allows the test to override the actual complexity of the implementation with mock versions, in a way that simply call two mock functions. If we have actual bugs in this logic, almost certainly it will be in the logic of these two functions, which remains untested, and not in the trivial lines inside the function that calls them. We're pretty much testing that AfterFunc works. Can we please make this whole logic normal?